File: compile.py
import os
import re
from urllib import parse
import time
import sys
from pathlib import Path

startTime = time.time()

# Get the current script's directory
current_dir = Path(__file__).parent

# Define the base folder dynamically based on the current directory
base_folder = current_dir / "src" / "client"

# Define paths for input and output files
init_file_path = base_folder / "init.client.lua"
output_file_path = base_folder / "full_code.lua"

# Function to escape multi-line strings for Luau
def escape_multiline_string(value):
    if "[[" in value or "]]" in value:
        delimiter_level = 1
        while f"[{'=' * delimiter_level}[" in value or f"]{'=' * delimiter_level}]" in value:
            delimiter_level += 1
        return f'[{"=" * delimiter_level}[{value}]{"=" * delimiter_level}]'
    else:
        return f'[[{value}]]'

# Function to convert a Python dictionary to a Luau table
def dict_to_luau_table(data, indent=0):
    lines = ["{"]
    indent_space = "    " * (indent + 1)

    for key, value in data.items():
        formatted_key = f'["{key}"]'
        if isinstance(value, dict):
            formatted_value = dict_to_luau_table(value, indent + 1)
        elif isinstance(value, list):
            formatted_value = list_to_luau_table(value, indent + 1)
        elif isinstance(value, str):
            if "\n" in value:
                formatted_value = escape_multiline_string(value)
            else:
                formatted_value = f'"{value}"'
        else:
            formatted_value = repr(value)
        lines.append(f'{indent_space}{formatted_key} = {formatted_value},')
    lines.append("    " * indent + "}")
    return "\n".join(lines)

# Function to convert a Python list to a Luau table
def list_to_luau_table(data, indent=0):
    indent_space = "    " * (indent + 1)
    lines = ["{"]
    for value in data:
        if isinstance(value, dict):
            formatted_value = dict_to_luau_table(value, indent + 1)
        elif isinstance(value, list):
            formatted_value = list_to_luau_table(value, indent + 1)
        elif isinstance(value, str):
            if "\n" in value:
                formatted_value = escape_multiline_string(value)
            else:
                formatted_value = f'"{value}"'
        else:
            formatted_value = repr(value)
        lines.append(f'{indent_space}{formatted_value},')
    lines.append("    " * indent + "}")
    return "\n".join(lines)

# Function to list Lua files and their content
def list_lua_files(base_folder):
    lua_files = {}
    ignore_list = ["full_code.lua", "init.client.lua"]
    for root, _, files in os.walk(base_folder):
        for file in files:
            relative_path = Path(root).joinpath(file).relative_to(base_folder)
            if file.endswith(".lua") and relative_path.name not in ignore_list:
                with open(base_folder / relative_path, "r", encoding="utf-8") as cur_file:
                    key = parse.quote(str(relative_path.with_suffix("")).replace("\\", "/"))
                    lua_files[key] = cur_file.read()
    return lua_files

# Read the init.client.lua file
with open(init_file_path, "r", encoding="utf-8") as file:
    data = file.read()

# Get Lua files as modules
modules = list_lua_files(base_folder)

# Replace preloaded module placeholder in the data
data = data.replace("C.preloadedModule = {}", "C.preloadedModule = " + dict_to_luau_table(modules))

# Write to the output file
with open(output_file_path, "w", encoding="utf-8") as write_file:
    write_file.write(data)

print(f"Successfully compiled in {time.time() - startTime:.2f} seconds")

File: compile.js
const fs = require('fs');
const path = require('path');
const url = require('url');

if (true){ return }

const startTime = Date.now();

// Get the current script's directory
const currentDir = __dirname;

// Define the base folder dynamically based on the current directory
const baseFolder = path.join(currentDir, 'src', 'client');

// Define paths for input and output files
const initFilePath = path.join(baseFolder, 'init.client.lua');
const outputFilePath = path.join(baseFolder, 'full_code.lua');

function sleep(ms) {
    return new Promise(resolve => setTimeout(resolve, ms));
}
// Function to escape multi-line strings for Luau
function escapeMultilineString(value) {
    if (value.includes('[[') || value.includes(']]')) {
        let delimiterLevel = 1;
        while (
            value.includes(`[${'='.repeat(delimiterLevel)}[`) ||
            value.includes(`]${'='.repeat(delimiterLevel)}]`)
        ) {
            delimiterLevel += 1;
        }
        return `[${'='.repeat(delimiterLevel)}[${value}]${'='.repeat(delimiterLevel)}]`;
    } else {
        return `[[${value}]]`;
    }
}

// Function to convert a JavaScript object to a Luau table
function dictToLuauTable(data, indent = 0) {
    const lines = ['{'];
    const indentSpace = '    '.repeat(indent + 1);

    for (const [key, value] of Object.entries(data)) {
        const formattedKey = `["${key}"]`;
        let formattedValue;

        if (typeof value === 'object' && !Array.isArray(value)) {
            formattedValue = dictToLuauTable(value, indent + 1);
        } else if (Array.isArray(value)) {
            formattedValue = listToLuauTable(value, indent + 1);
        } else if (typeof value === 'string') {
            formattedValue = value.includes('\n')
                ? escapeMultilineString(value)
                : `"${value}"`;
        } else {
            formattedValue = JSON.stringify(value);
        }

        lines.push(`${indentSpace}${formattedKey} = ${formattedValue},`);
    }

    lines.push('    '.repeat(indent) + '}');
    return lines.join('\n');
}

// Function to convert a JavaScript array to a Luau table
function listToLuauTable(data, indent = 0) {
    const indentSpace = '    '.repeat(indent + 1);
    const lines = ['{'];

    for (const value of data) {
        let formattedValue;

        if (typeof value === 'object' && !Array.isArray(value)) {
            formattedValue = dictToLuauTable(value, indent + 1);
        } else if (Array.isArray(value)) {
            formattedValue = listToLuauTable(value, indent + 1);
        } else if (typeof value === 'string') {
            formattedValue = value.includes('\n')
                ? escapeMultilineString(value)
                : `"${value}"`;
        } else {
            formattedValue = JSON.stringify(value);
        }

        lines.push(`${indentSpace}${formattedValue},`);
    }

    lines.push('    '.repeat(indent) + '}');
    return lines.join('\n');
}

// Function to list Lua files and their content
function listLuaFiles(baseFolder) {
    const luaFiles = {};
    const ignoreList = ['full_code.lua', 'init.client.lua'];

    const traverseDirectory = (dir) => {
        fs.readdirSync(dir, { withFileTypes: true }).forEach((entry) => {
            const entryPath = path.join(dir, entry.name);
            if (entry.isDirectory()) {
                traverseDirectory(entryPath);
            } else if (
                entry.isFile() &&
                entry.name.endsWith('.lua') &&
                !ignoreList.includes(entry.name)
            ) {
                // Get the relative path and convert it to a forward-slash format
                const relativePath = path.relative(baseFolder, entryPath).replace(/\\/g, '/');
                const key = relativePath.replace(/\.lua$/, ''); // Remove the .lua extension
                luaFiles[key] = fs.readFileSync(entryPath, 'utf-8');
            }
        });
    };

    traverseDirectory(baseFolder);
    return luaFiles;
}

// Read the init.client.lua file
const data = fs.readFileSync(initFilePath, 'utf-8');

// Get Lua files as modules
const modules = listLuaFiles(baseFolder);

// Replace preloaded module placeholder in the data
const replacedData = data.replace(
    'C.preloadedModule = {}',
    `C.preloadedModule = ${dictToLuauTable(modules)}`
);

// Write to the output file
fs.writeFileSync(outputFilePath, replacedData, 'utf-8');

console.log(`Successfully compiled in ${(Date.now() - startTime) / 1000} seconds`);
